pkg/operator/ceph/version/version_test.go

```diff
@@ -0,0 +1,114 @@
+package version
+
+import (
+	"fmt"
+	"github.com/stretchr/testify/assert"
+	"testing"
+)
+
+func TestToString(t *testing.T) {
+	assert.Equal(t, "14.0.0 nautilus", fmt.Sprintf("%s", &Nautilus))
+	assert.Equal(t, "13.0.0 mimic", fmt.Sprintf("%s", &Mimic))
+	assert.Equal(t, "12.0.0 luminous", fmt.Sprintf("%s", &Luminous))
+
+	expected := fmt.Sprintf("-1.0.0 %s", unknownVersionString)
+	assert.Equal(t, expected, fmt.Sprintf("%s", &CephVersion{-1, 0, 0}))
+}
+
+func TestReleaseName(t *testing.T) {
+	assert.Equal(t, "nautilus", Nautilus.ReleaseName())
+	assert.Equal(t, "mimic", Mimic.ReleaseName())
+	assert.Equal(t, "luminous", Luminous.ReleaseName())
+	ver := CephVersion{-1, 0, 0}
+	assert.Equal(t, unknownVersionString, ver.ReleaseName())
+}
+
+func extractVersionHelper(t *testing.T, text string, major, minor, extra int) {
+	v, err := ExtractCephVersion(text)
+	if assert.NoError(t, err) {
+		assert.Equal(t, v, CephVersion{major, minor, extra})
+	}
+}
+
+func TestExtractVersion(t *testing.T) {
+	// release build
+	v0c := "ceph version 12.2.8 (ae699615bac534ea496ee965ac6192cb7e0e07c0) luminous (stable)"
+	v0d := `
+root@7a97f5a78bc6:/# ceph --version
+ceph version 12.2.8 (ae699615bac534ea496ee965ac6192cb7e0e07c0) luminous (stable)
+`
+	extractVersionHelper(t, v0c, 12, 2, 8)
+	extractVersionHelper(t, v0d, 12, 2, 8)
+
+	// development build
+	v1c := "ceph version 14.1.33-403-g7ba6bece41"
+	v1d := `
+[nwatkins@smash build]$ bin/ceph --version
+*** DEVELOPER MODE: setting PATH, PYTHONPATH and LD_LIBRARY_PATH ***
+ceph version 14.1.33-403-g7ba6bece41
+(7ba6bece4187eda5d05a9b84211fe6ba8dd287bd) nautilus (rc)
+`
+	extractVersionHelper(t, v1c, 14, 1, 33)
+	extractVersionHelper(t, v1d, 14, 1, 33)
+
+	// build without git version info
+	v2c := "ceph version Development (no_version) nautilus (rc)"
+	v2d := `
+[nwatkins@daq build]$ bin/ceph --version
+*** DEVELOPER MODE: setting PATH, PYTHONPATH and LD_LIBRARY_PATH ***
+ceph version Development (no_version) nautilus (rc)
```

Comment:
Is this what the master image returns? How do we expect rook to behave in this case? I would hope we could at least detect "nautilus" and set the major version to 14. Otherwise, we'll get unpredictable behavior during orchestration. Or will the orchestration fail with this version check?

Reply:
This is what is produced on special builds in which version information is explicitly turned off. Normal master builds always include the X.Y.Z version information. I think this case can be removed, actually---it's kind of confusing. AFAIK the only reason to turn of the version information is to make Ceph builds quicker.

Reply:
Just for some clarifying context, I saw this post on ceph-devel today

```
- I will periodically merge nautilus back into master, up to and including 
the final/first stable v14.2.0 nautilus release.  (This is to ensure the 
master 'git describe' version is >= nautilus, although shortly after 
nautilus is out we'll also create a v15.0.0 tag just to kick off the 
octopus cycle.)
```

Reply:
I agree with Travis' first thought here. I think we need to consider the case of developers working on Ceph's code base. Speeding up Ceph builds is good in a dev context for faster iteration, and if the X.Y.Z version info isn't important and speeds up a build, we can expect devs will want to omit that information.